// firebase-tools/lib/emulator/auth/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthEmulator = void 0;
const utils = require("../../utils");
const constants_1 = require("../constants");
const types_1 = require("../types");
const server_1 = require("./server");
class AuthEmulator {
    constructor(args) {
        this.args = args;
    }
    async start() {
        const { host, port } = this.getInfo();
        const app = await server_1.createApp(this.args.projectId);
        const server = app.listen(port, host);
        this.destroyServer = utils.createDestroyer(server);
    }
    async connect() {
    }
    stop() {
        return this.destroyServer ? this.destroyServer() : Promise.resolve();
    }
    getInfo() {
        const host = this.args.host || constants_1.Constants.getDefaultHost(types_1.Emulators.AUTH);
        const port = this.args.port || constants_1.Constants.getDefaultPort(types_1.Emulators.AUTH);
        return {
            name: this.getName(),
            host,
            port,
        };
    }
    getName() {
        return types_1.Emulators.AUTH;
    }
}
exports.AuthEmulator = AuthEmulator;
